### PR TITLE
Add NN SKU token provider

### PR DESCRIPTION
**Pull request by Guardiola31337**
## Description

Adds NN SKU token provider

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

All requests coming from the SDK should append a SKU token, including ART which is done internally from NN

### Implementation

Implement `SkuToken` callback so that NN can retrieve the current SKU token from the SDK and ensures that `Navigator#setSkuTokenSource` is called before `Navigator#configureRouter`

TODO

- [x] Replace `ms-free-drive-billing-SNAPSHOT-4` by a NN stable version
  - [x] ~`12.0.0`~
  - [x] ~`12.0.1`~
  - [x] `12.0.2`

## Testing

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

cc @asinghal22 @mskurydin

### CODE REVIEW

**Comment by asinghal22** [comment]
See attached document; Please know I don't know how difficult it is to create unit tests so feel free to cherry pick what you deem appropriate.

https://docs.google.com/document/d/1OXG4hY09LS-pxzp6gC-p6fzP3iXhRR0gQ4o4zLRxIwE/edit#

File: libdirections-onboard/src/test/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouterTest.kt [hunk 1]
```diff
@@ -61,11 +62,13 @@ class MapboxOnboardRouterTest {
     private val routerResultFailure: RouterResult = mockk(relaxUnitFun = true)
     private val routerOptions: RouteOptions = provideDefaultRouteOptions()
     private val logger: Logger = mockk(relaxUnitFun = true)
+    private val mockSkuTokenProvider = mockk<SkuTokenProvider>(relaxed = true)
 
     @Before
     fun setUp() {
         every { navigator.configureRouter(any(), any()) } returns 0
-        onboardRouter = MapboxOnboardRouter(ACCESS_TOKEN, navigator, MapboxOnboardRouterConfig(TILE_PATH), logger)
+        every { mockSkuTokenProvider.obtainSkuToken() } returns ("102ka34odzf38e3b8f5f1ba42818e94d31090d6479f")
```
Comment (kmadsen) [inline libdirections-onboard/src/test/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouterTest.kt:70]: what happens when obtainSkuToken is empty? silent fails?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt [hunk 2]
```diff
@@ -71,6 +73,16 @@ class MapboxNavigationAccounts private constructor() : SkuTokenProvider {
         } ?: resourceUrl
     }
 
+    /**
+     * Returns current SDK SKU token needed for ART.
+     */
+    override fun obtainSkuToken(): String {
+        return skuGenerator?.let { generator ->
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:80]: maybe :)

return skuGenerator?.also { generateToken() } ?: EMPTY_SKU

File: libdirections-onboard/src/test/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouterTest.kt (hunk 1 above)
Comment (Guardiola31337) [reply libdirections-onboard/src/test/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouterTest.kt:70]: It's handled downstream by NN but to answer your question the token it's not appended to the URL.

Comment (mskurydin) [reply libdirections-onboard/src/test/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouterTest.kt:70]: if there is no token source, we shall just use the regular URL.
The same happens if it returns an empty token.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt (hunk 2 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:80]: The compiler doesn't like it 😬 

obtainSkuToken suggestion

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:80]: oh yeah, also and apply return SkuGenerator

you can just 
```
return skuGenerator?.generateToken() ?: EMPTY_SKU
```

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:80]: Even better

```kotlin
override fun obtainSkuToken(): String = skuGenerator?.generateToken() ?: EMPTY_SKU
```

Thanks for the suggestion!

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NativeSkuTokenProvider.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package com.mapbox.services.android.navigation.v5.navigation
+
+import com.mapbox.navigator.SkuTokenSource
+
+internal class NativeSkuTokenProvider : SkuTokenSource() {
+    override fun getToken(): String = ""
+}
```
Comment (abhishek1508) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NativeSkuTokenProvider.kt:7]: Are we adding this new file to `Legacy` SDK? 🤔

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NativeSkuTokenProvider.kt:7]: Yeah so everything keeps working in the legacy. In any case, we don't support SKU in the legacy so this should do the trick 👀 https://github.com/mapbox/mapbox-navigation-android/pull/2955#discussion_r424601044

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt [hunk 4]
```diff
@@ -71,6 +73,11 @@ class MapboxNavigationAccounts private constructor() : SkuTokenProvider {
         } ?: resourceUrl
     }
 
+    /**
+     * Returns current SDK SKU token needed for ART.
+     */
+    override fun obtainSkuToken(): String = skuGenerator?.generateToken() ?: EMPTY_SKU
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:79]: Does this need to be `@Synchronized` 🤔 @kmadsen @abhishek1508

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt [hunk 5]
```diff
@@ -71,6 +73,11 @@ class MapboxNavigationAccounts private constructor() : SkuTokenProvider {
         } ?: resourceUrl
     }
 
+    /**
+     * Returns current SDK SKU token needed for ART.
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:77]: Basic question, what's ART? Can we spell it out for the comment?

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:77]: Api Routing Tiles

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt (hunk 4 above)
Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/accounts/MapboxNavigationAccounts.kt:79]: if it does, would prefer the callback approach over blocking

**Comment by Guardiola31337** [comment]
Going ahead and merging here 🚀

**Pull request merged**